nomos-services/data-availability/indexer/src/lib.rs

```diff
@@ -0,0 +1,420 @@
+pub mod consensus;
+pub mod indexer;
+pub mod storage;
+
+use std::fmt::{Debug, Formatter};
+use std::hash::Hash;
+use std::sync::mpsc::Sender;
+
+use consensus::ConsensusAdapter;
+use cryptarchia_consensus::network::NetworkAdapter;
+use cryptarchia_consensus::CryptarchiaConsensus;
+use futures::StreamExt;
+use indexer::DaIndexer;
+use nomos_core::block::Block;
+use nomos_core::da::certificate::{BlobCertificateSelect, Certificate};
+use nomos_core::header::HeaderId;
+use nomos_core::tx::{Transaction, TxSelect};
+use nomos_mempool::da::verify::fullreplication::DaVerificationProvider;
+use nomos_mempool::verify::MempoolVerificationProvider;
+use nomos_mempool::{backend::MemPool, network::NetworkAdapter as MempoolAdapter};
+use nomos_storage::backends::StorageBackend;
+use nomos_storage::StorageService;
+use overwatch_rs::services::handle::ServiceStateHandle;
+use overwatch_rs::services::life_cycle::LifecycleMessage;
+use overwatch_rs::services::relay::{Relay, RelayMessage};
+use overwatch_rs::services::state::{NoOperator, NoState};
+use overwatch_rs::services::{ServiceCore, ServiceData, ServiceId};
+use overwatch_rs::DynError;
+use serde::de::DeserializeOwned;
+use serde::{Deserialize, Serialize};
+use storage::DaStorageAdapter;
+use tracing::error;
+
+pub type ConsensusRelay<
+    A,
+    ClPool,
+    ClPoolAdapter,
+    DaPool,
+    DaPoolAdapter,
+    DaVerificationProvider,
+    TxS,
+    BS,
+    Storage,
+> = Relay<
+    CryptarchiaConsensus<
+        A,
+        ClPool,
+        ClPoolAdapter,
+        DaPool,
+        DaPoolAdapter,
+        DaVerificationProvider,
+        TxS,
+        BS,
+        Storage,
+    >,
+>;
+
+pub struct DataIndexerService<
+    Indexer,
+    DaStorage,
+    Consensus,
+    A,
+    ClPool,
+    ClPoolAdapter,
+    DaPool,
+    DaPoolAdapter,
+    DaVerificationProvider,
+    TxS,
+    BS,
+    ConsensusStorage,
+> where
+    Indexer: DaIndexer,
+    Indexer::Blob: 'static,
+    Indexer::VID: 'static,
+    A: NetworkAdapter,
+    ClPoolAdapter: MempoolAdapter<Payload = ClPool::Item, Key = ClPool::Key>,
+    ClPool: MemPool<BlockId = HeaderId>,
+    DaPool: MemPool<BlockId = HeaderId>,
+    DaPoolAdapter: MempoolAdapter<Key = DaPool::Key>,
+    DaPoolAdapter::Payload: Certificate + Into<DaPool::Item> + Debug,
+    DaVerificationProvider: MempoolVerificationProvider<
+        Payload = DaPoolAdapter::Payload,
+        Parameters = <DaPoolAdapter::Payload as Certificate>::VerificationParameters,
+    >,
+    ClPool::Item: Clone + Eq + Hash + Debug + 'static,
+    ClPool::Key: Debug + 'static,
+    DaPool::Item: Clone + Eq + Hash + Debug + 'static,
+    DaPool::Key: Debug + 'static,
+    A::Backend: 'static,
+    TxS: TxSelect<Tx = ClPool::Item>,
+    BS: BlobCertificateSelect<Certificate = DaPool::Item>,
+    DaStorage: DaStorageAdapter,
+    ConsensusStorage: StorageBackend + Send + Sync + 'static,
+{
```

Comment:
Do we need the bounds in the struct declaration?

Reply:
These bounds are required mainly for a consensus relay that is used to subscribe and receive blocks. What we probably want is something between `ServiceData` and `CryptarchiaConsensus`, like `NomosConsensus`, which could be generic over block and its items?

Reply:
> These bounds are required mainly for a consensus relay that is used to subscribe and receive blocks. What we probably want is something between `ServiceData` and `CryptarchiaConsensus`, like `NomosConsensus`, which could be generic over block and its items?

I just wasnt sure if just adding the generics (without the bounds) would be enough. As usually declarations can be just that. But yeah, if we have to add all of them. Maybe we can bound them excep for the block and its items. Or as the work is done already just left it as it is right now. No problem.

Reply:
I fiddled with these definitions, unfortunately Overwatch is pretty strict when trying to fetch a handle to another service.

Reply:
Yeah, is the price to pay for having everything checked on compilation time.